Migrate `left_sidebar` and `right_sidebar` to handlebars.

aryanshridhar opened the pull request:
Added a couple of commits to migrate the `left_sidebar.html` (8ba6d0a) and `right_sidebar.html` (9c28c91) to handlebars.
The templates (`left_sidebar.hbs` and `right_sidebar.hbs`) are then rendered using `ui_init.js`. 

Fixes parts of #18792.

aryanshridhar commented on zerver/tests/test_home.py:
```diff
@@ -704,11 +704,11 @@ def test_invites_by_admins_only(self) -> None:
 
         self.login_user(user_profile)
         self.assertFalse(user_profile.is_realm_admin)
-        self.assertNotInHomePage("Invite more users")
```
`assertNotInHomePage` and `assertInHomePage` basically checks for a string that is either absent or present in the `index.html` template. However, since the `right_sidebar` is migrated to handlebars, `assertInHomePage` will always return `false`.
The way I modified the tests are so that they `assertEqual` if the user_profile invite policy matches the `page_params.show_invites` (Because this is what we check within handlebars). 

Any suggestions regarding this would help :)

ganpa3 replied:
I think the purpose of this function is to check if a user can invite others, which is already tested by `test_compute_show_invites_and_add_streams*` functions

timabbott replied:
Yeah, agreed, I think we can just delete this entire test, perhaps in a commit message explaining that the behavior is already tested by the other logic for deciding what to pass in `page_params`.

(We may post-migration want to add new tests in the Handlebars world)

aryanshridhar replied:
Yup, Thanks @ganpa3 @timabbott!

aryanshridhar commented:
Updated the pr along with its commit message.
Would be great if @timabbott can have a look at it again :)

aryanshridhar commented:
Latest push changes along with rebase - 
 * Removed redundant backend tests in `test_home` due to the template migration.
 * Updated commit message to reflect the same changes.

timabbott commented on static/templates/left_sidebar.hbs:
```diff
@@ -42,46 +42,46 @@
                     </span>
-                    {#- squash whitespace -#}
-                    <span>{{ _('Recent topics') }}</span>
+                    {{!-- squash whitespace --}}
```
The `squash whitespace` comments were not correctly migrated to actually squash whitespace using handlebars whitespace control; I fixed that while merging.

timabbott commented on templates/zerver/app/index.html:
```diff
@@ -91,7 +91,7 @@ <h3>{{ _('Loading...') }}</h3>
     <div class="app-main">
         <div class="column-left">
-            {% include "zerver/app/left_sidebar.html" %}
+            <div id = "left-sidebar-holder"></div>
```
I decided it was better to add `left-sidebar-container` as an ID on the `column-left` class here, to avoid creating an additional unnecessary div.

timabbott replied:
Also note that we don't use space around the `=` in our HTML style.

timabbott commented on static/js/ui_init.js:
```diff
@@ -464,6 +473,13 @@ export function initialize_everything() {
+    // to build the page layout early so that it's rendered content
+    // can be used by other modules.
+    // For example- Populating the stream lists/populating private messages.
+    initialize_left_sidebar();
```
I rewrote the comments, since this should be next to `compose` and share its comment.

timabbott commented on static/js/ui_init.js:
```diff
@@ -138,6 +139,14 @@ function initialize_left_sidebar() {
 
+function initialize_right_sidebar() {
+    const rendered_sidebar = render_right_sidebar({
+        can_invite_others_to_realm: page_params.show_invites,
```
Changed this to `settings_data.user_can_invite_others_to_realm()`, since we can compute this on the frontend.

This may mean we get to remove `compute_show_invites` as well in a follow-up PR.

timabbott commented:
Merged after making that batch of changes, thanks @aryanshridhar!  

Can you look at whether we get to delete the backend `can_show_invites` function as well?  I think the answer is yes.
